etna/lib/etna/directed_graph.rb

```diff
@@ -19,17 +19,38 @@ def add_connection(parent, child)
   end
 
   def descendants(parent)
-    result = []
     seen = Set.new
-    queue = @children[parent].keys
+
+    seen.add(parent)
```

Comment:
I didn't end up actually using this directed_graph implementation with the simplified approach to restricted recursion, but I thought it was still useful so I left the code.  If you feel this is distracting and not useful, I'll happily remove.